freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py

```diff
@@ -167,3 +180,149 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
         )
         images = lb.get_images_by_nvrs(rebuild_images_list, rpm_nvrs=errata_rpm_nvrs)
         return images
+
+    def _get_module_name_stream_version_set(self):
+        """
+        Get set of modules with format NAME:STREAM:VERSION in an advisory.
+
+        :return: a set with item NAME:STREAM:VERSION
+        :rtype: set
+        """
+        module_name_stream_version_set = set()
+        with koji_service(
+            conf.koji_profile, log, login=False, dry_run=self.dry_run
+        ) as session:
+            for advisory_module_nvr in self.advisory_module_nvrs:
+                mmd = session.get_modulemd(advisory_module_nvr)
+                name = mmd.get_module_name()
+                stream = mmd.get_stream_name()
+                version = mmd.get_version()
+                module_name_stream_version_set.add(
+                    ":".join([name, stream, str(version)])
+                )
+
+        return module_name_stream_version_set
+
+    def _prepare_data_for_compose(
+        self, original_odcs_compose_ids, module_name_stream_version_set
+    ):
+        """
+        Prepare data before recording compose data in build or
+        creating a new compose and linking it to a build.
+
+        :param original_odcs_compose_ids list: A list of compose ids
+            in a image.
+        :param module_name_stream_version_set set: A set of module's metadata
+            in format NAME:STREAM:VERSION of an advisory.
+        :return: a tuple of compose data. The first item is a set of
+            compose ids which will be renewed in build, the second item
+            is a set of required data for creating new composes
+        :rtype: tuple
+        """
+        renewed_odcs_compose_set = set()
+        new_odcs_compose_set = set()
+        for compose_id in original_odcs_compose_ids:
+            compose = self.odcs.get_compose(compose_id)
+            source_type = compose.get("source_type")
+            if source_type != PungiSourceType.MODULE:
+                renewed_odcs_compose_set.add(compose_id)
+                continue
+
+            name_stream_set = set()
+            source_value = compose.get("source", "")
+            for source in source_value.split():
+                group = source.strip().split(":")[:3]
+                name_stream_set.add(":".join(group))
+
+            # If all the sources are not in advisory, renew them
+            if len(name_stream_set - module_name_stream_version_set) == len(
+                name_stream_set
+            ):
+                renewed_odcs_compose_set.add(compose_id)
+            else:
+                # If some source in advisory, create composes for all separately
+                new_odcs_compose_set.update(name_stream_set)
```

Comment:
So when some sources are in advisory, you'll create new composes for all these old modules, right? I believe this is wrong, for example, there is old modules:

- module-a:stable:123
- module-b:el8:345

And now you have module-b:el8:350 in advisory, you should request 2 new module composes (in case you want to create one compose for each module):
- compose: module-a:stable:123
- compose: module-b:el8:350

compose for `module-b:el8:345` should be excluded, in your code, there will be 3 composes with all these modules included.

Reply:
Ack. Fixed